pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3722,8 +3722,12 @@ public String startReplaceSegments(String tableNameWithType, List<String> segmen
     // Trigger the proactive segment clean up if needed. Once the lineage is updated in the property store, it
     // is safe to physically delete segments.
     if (!segmentsToCleanUp.isEmpty()) {
-      LOGGER.info("Cleaning up the segments while startReplaceSegments: {}", segmentsToCleanUp);
-      deleteSegments(tableNameWithType, segmentsToCleanUp);
+      // Only keep the segments that within the table for the proactive clean-up.
+      segmentsToCleanUp.retainAll(getSegmentsFor(tableNameWithType, false));
```

Comment:
We also nee to remove Zk metadata and from IS state ? hopefully getSegmentsFor returns IS segments as well.

Reply:
this only keeps the segments not deleted from the idealStates, so we won't do unnecessary clean up.

Reply:
I don't follow this. Why do we want to skip segments not in IS? One possible scenario is that segments are not yet pushed to IS, and we don't want to leave them orphan.